src-cpp/master/master.cc

```diff
@@ -12,24 +12,100 @@
 using namespace std;
 using namespace asio::ip;
 
-Master::Master(string& filename,vector<udp::endpoint>& r_endpts) :
-  remote_endpts(r_endpts)
+Master::Master(string& fname,vector<udp::endpoint>& r_endpts) :
+  remote_endpts(r_endpts), io_service(), socket(io_service,udp::endpoint(udp::v4(),0)) 
 {
-  msockets.reserve(r_endpts.size());
-  for (auto endpt : remote_endpts) {
-    msockets.push_back(new MSocket(io_service,endpt,filename));
-  }
+  file = SndfileHandle (fname);
+
+  std::cerr << "Opened file '" << fname << "'" << std::endl;
+  std::cerr << "\tSample rate '" << file.samplerate() << "'" << std::endl;
+  std::cerr << "\tChannels '" << file.channels() << "'" << std::endl;
 }
 
 Master::~Master(){
-  for (MSocket *sock : msockets) {
-    delete sock;
+}
+
+void Master::send_timesync(udp::endpoint& remote_endpt) {
+  TPacket tp;
+  tp.from_sent = get_millisecond_time();
+
+  socket.async_send_to(asio::buffer(tp.pack()),remote_endpt,
+      [this,remote_endpt](error_code,size_t) {
+          this->receive_timesync_reply(remote_endpt);
+      });
+}
+
+// send a timesync to every remote endpoint.
+void Master::send_timesync(){
+  for (udp::endpoint& remote_endpt : remote_endpts) {
+    send_timesync(remote_endpt);
   }
 }
 
-void Master::run(){
-  for (MSocket *sock : msockets) {
-    sock->start();
+void Master::receive_timesync_reply(udp::endpoint remote_endpt) {
+  socket.async_receive_from(
+      asio::buffer(this->tp_buffer, TP_BUFFER_SIZE), remote_endpt,
+      [this,remote_endpt](error_code e, size_t bytes_recvd) {
+        // calculate sum shit
+
+        // immediately grab the receipt time  
+        mtime_t from_recv = get_millisecond_time();
+        
+        // unpack the time packet
+        Packet * p = Packet::unpack(this->tp_buffer, bytes_recvd);
+        TPacket * tp = static_cast<TPacket *> (p);
+        tp->from_recvd = from_recv;
+        tp->tp_type = COMPLETE;
+
+        // calculate the offset
+        mtime_offset_t offset = ((tp->to_recvd - tp->from_sent) + (tp->to_sent - tp->from_recvd))/2;
+        tp->offset = offset;
+
+        // and send the reply
+        socket.async_send_to(
+            asio::buffer(tp->pack()), remote_endpt,
+            [this](error_code, size_t) {
+              // reply sent...
+              // do we need timeouts on this shit
+            }
+        );
+        // ready to start sending data
+        this->send_data();
```

Comment:
Should we just wait until all remote endpoints have been synced before sending data?  Won't this do so after receiving the first client's reply?